global/scripts/parallel.shlib

```diff
@@ -172,3 +185,73 @@ function par_addjob_stdin()
     par_jobinputarray+=("$jobinput")
 }
 
+#Add job to the current stage (FSLSUB, BUILTIN or NONE mode)
+#Inputs:
+#1. parallel mode
+#2. queue name for FSLSUB (local submit if empty)
+#3. command with arguments
+function par_add_job_to_stage()
+{
+    local PARALLEL_MODE=$1
+    shift
+
+    #FSLSUB queue name for FSLSUB mode, should be 'NONE' for other modes.
+    local fslsub_options fslsub_queue=$1
+    shift
+
+    local job submit_cmd cmd=("$@")
+
+    if [ "$PARALLEL_MODE" == "FSLSUB" ]; then #current job is submitted, and added to fsl_job_list
+        fslsub_options=""
+        submit_cmd=(fsl_sub)
+        if [[ -n "$fslsub_queue" ]]; then 
+            submit_cmd+=(-q "$fslsub_queue")
+        fi
+        if [[ -n "$par_fslsub_job_waitlist" ]]; then
+            submit_cmd+=(-j "$par_fslsub_job_waitlist")
+        fi
+        submit_cmd+=("${cmd[@]}")        
+        job=$("${submit_cmd[@]}")
+        par_fslsub_job_list+=("$job")
+    elif [ "$PARALLEL_MODE" == "BUILTIN" ]; then          #current job is added to the list of local parallel jobs        
+        par_addjob $cmd        
+    else #NONE                                          #current job is executed.
+        echo "launching, parallel mode=NONE, will wait to finish: ${cmd[*]}"
+        ${cmd} 1> ./"$par_global_scriptname.$$.$par_global_launch_count.o.log" 2> ./"$par_global_scriptname.$$.$par_global_launch_count.e.log"
+        par_global_launch_count=$((par_global_launch_count + 1))        
+    fi
+}
+
+#Finalize the list of jobs at the current stage.
+#FSLSUB: update prerequisite job waitlist for the next stage; empty current stage job list
+#BUILTIN: wait until all jobs at the current stage complete, internally, also empty current job list
+#NONE: do nothing, execution was done in par_add_job_to_stage_multimode()
+
+#Inputs:
+#1. Mode name
+#2. max jobs (BUILTIN mode only)
+function par_finalize_stage()
+{
+    #global fslsub_job_list fslsub_job_waitlist
+    local PARALLEL_MODE=$1
+    shift    
+    local temp max_jobs=$1
+    if [ -z "$max_jobs" ]; then max_jobs=-1; fi
+
+    if [ "$PARALLEL_MODE" == "FSLSUB" ]; then           #update the fslsub_job_waitlist with the current job list.
+        par_fslsub_job_waitlist=""
+        if [ -n "$par_fslsub_job_list" ]; then 
+            temp="${par_fslsub_job_list[*]}"
+            par_fslsub_job_waitlist="${temp// /,}"
+        fi
+        par_fslsub_job_list=() 
+        echo "updating job waitlist for fsl_sub done"
```

Comment:
Please remove this low-information echo, since all that this code is doing is swapping variables around, it doesn't need progress indication messages.

I suppose the echos in the BUILTIN section below are okay, since there will be a meaningful delay between them while it waits for the jobs.

Reply:
Ok